content/docs/userspace/gall-guide/8-subscriptions.md

```diff
@@ -0,0 +1,875 @@
+
+The example will be composed of two separate agents - a publisher called
+`/app/todo.hoon` and a subscriber called `/app/todo-watcher.hoon`, which will
+live on separate ships. It will be a very rudimentary To-Do app - to-do tasks
+will be poked into the publisher and sent out to the subscriber as `%fact`s,
+which will just print them to the dojo. It will have its types defined in
+`/sur/todo.hoon`, and it will have a couple of `mark` files for pokes and
+updates: `/mar/todo/action.hoon` and `/mar/todo/update.hoon`.
+
+Before we get into trying it out, we'll first walk through the `/sur` file, mark
+files, and each agent.
+
+### Types and marks
+
+#### `/sur/todo.hoon`
+
+```hoon
+|%
++$  id  @
++$  name  @t
++$  task  [=name done=?]
++$  tasks  (map id task)
++$  who  @p
++$  friends  (set who)
++$  action
+  $%  [%add =name]
+      [%del =id]
+      [%toggle =id]
+      [%rename =id =name]
+      [%allow =who]
+      [%kick =who]
+  ==
++$  update
+  $%  [%add =id =name]
+      [%del =id]
+      [%toggle =id]
+      [%rename =id =name]
+      [%initial =tasks]
+  ==
+--
+```
+
+This file defines most of the types for the agents. The list of to-do tasks will
+be stored in the state of the publisher agent as the `tasks` type, a `(map id task)`, where a `task` is a `[=name done=?]`. The set of ships allowed to
+subscribe will be stored in `friends`, a `(set @p)`, also in the publisher's
+state. After that, there are the head-tagged unions of accepted poke `action`s
+and `update`s for subscribers.
+
+#### `/mar/todo/action.hoon`
+
+```hoon
+/-  todo
+|_  =action:todo
+++  grab
+  |%
+  ++  noun  action:todo
+  --
+++  grow
+  |%
+  ++  noun  action
+  --
+++  grad  %noun
+--
+```
+
+This is a very simple mark file for the `action` type.
+
+#### `/mar/todo/update.hoon`
+
+```hoon
+/-  todo
+|_  =update:todo
+++  grab
+  |%
+  ++  noun  update:todo
+  --
+++  grow
+  |%
+  ++  noun  update
+  --
+++  grad  %noun
+--
+```
+
+This is a very simple mark file for the `update` type.
+
+### Publisher
+
+#### `/app/todo.hoon`
+
+```hoon
+/-  todo
+/+  default-agent, dbug
+|%
++$  versioned-state
+  $%  state-0
+  ==
++$  state-0  [%0 =friends:todo =tasks:todo]
++$  card  card:agent:gall
+--
+%-  agent:dbug
+=|  state-0
+=*  state  -
+^-  agent:gall
+|_  =bowl:gall
++*  this  .
+    def   ~(. (default-agent this %.n) bowl)
+++  on-init
+  ^-  (quip card _this)
+  `this
+++  on-save
+  ^-  vase
+  !>(state)
+++  on-load
+  |=  old-state=vase
+  ^-  (quip card _this)
+  =/  old  !<(versioned-state old-state)
+  ?-  -.old
+    %0  `this(state old)
+  ==
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  |^
+  ?>  =(src.bowl our.bowl)
+  ?+    mark  (on-poke:def mark vase)
+      %todo-action
+    =^  cards  state
+      (handle-poke !<(action:todo vase))
+    [cards this]
+  ==
+  ++  handle-poke
+    |=  =action:todo
+    ^-  (quip card _state)
+    ?-    -.action
+        %add
+      :_  state(tasks (~(put by tasks) now.bowl [name.action %.n]))
```

Comment:
how can it get more than one in one event?

Reply:
Some local agent might send multiple `%poke`s to this app, which then get processed immediately, within the same event, because these all flow from the one external event that triggered the local agent to send the `%poke`s. As such the `now.bowl` will be the same for both those `+on-poke` invocations.

Reply:
Oh! That is something I was puzzling over last week. I was cribbing off of this example for my own app and couldn't get more than one `%poke` to go through at once ended up wasting a couple hours and then not understanding how I fixed it.

Reply:
lol interesting I didn't consider this case. I might add an explanation of exactly when the bowl gets repopulated when that's introduced. I'm unsure about having them fix this as an exercise - I might just add a small recursion that checks for existence and adds 1ms to the time if it's already in the map, then tries again. I can then explain why I did that in the code walkthrough, harkening back to the lesson when the bowl was introduced